pkg/kubelet/network/kubenet/kubenet_linux.go

```diff
@@ -608,6 +629,69 @@ func (plugin *kubenetNetworkPlugin) getActivePods() ([]*hostport.ActivePod, erro
 	return activePods, nil
 }
 
+// ipamGarbageCollection will release unused IP.
+func (plugin *kubenetNetworkPlugin) ipamGarbageCollection() {
+	glog.V(2).Infof("Start IP garbage collection")
+
+	ipamDir := filepath.Join(defaultIPAMDir, KubenetPluginName)
+	files, err := ioutil.ReadDir(ipamDir)
+	if err != nil {
+		glog.Errorf("Failed to list files in %q: %v", ipamDir, err)
+		return
+	}
+
+	// gather containerIDs for allocated ips
+	ipContainerIdMap := make(map[string]string)
+	for _, file := range files {
+		// skip non checkpoint file
+		if ip := net.ParseIP(file.Name()); ip == nil {
+			continue
+		}
+
+		content, err := ioutil.ReadFile(filepath.Join(ipamDir, file.Name()))
+		if err != nil {
+			glog.Errorf("Failed to read file %v: %v", file, err)
+		}
+		ipContainerIdMap[file.Name()] = strings.TrimSpace(string(content))
+	}
+
+	// gather infra container IDs of current running Pods
+	runningContainerIDs := utilsets.String{}
+	pods, err := plugin.getNonExitedPods()
+	if err != nil {
+		glog.Errorf("Failed to get pods: %v", err)
+		return
+	}
+	for _, pod := range pods {
+		containerID, err := plugin.host.GetRuntime().GetPodContainerID(pod)
+		if err != nil {
+			glog.Errorf("Failed to get infra containerID of %q/%q: %v", pod.Namespace, pod.Name, err)
+			continue
+		}
+
+		runningContainerIDs.Insert(strings.TrimSpace(containerID.ID))
+	}
+
+	// release leaked ips
+	for ip, containerID := range ipContainerIdMap {
+		// if the container is not running, release IP
+		if !runningContainerIDs.Has(containerID) {
+			// CNI requires all config to be presented, although only containerID is needed in this case
+			rt := &libcni.RuntimeConf{
+				ContainerID: containerID,
+				IfName:      network.DefaultInterfaceName,
+				NetNS:       "",
+			}
+
+			glog.V(2).Infof("Releasing IP %q allocated to %q.", ip, containerID)
+			// CNI bridge plugin should try to release IP and then return
+			if err := plugin.cniConfig.DelNetwork(plugin.netConfig, rt); err != nil {
```

Comment:
I was asking if there's a race with this thing deleting an ip used by a hash it thinks is unused, and another thread allocating an ip for that hash, or something along those lines